src/transformers/models/iris/modeling_iris.py

```diff
@@ -0,0 +1,1964 @@
+        offsets = torch.arange(max_blocks).repeat_interleave(self.num_kept_tokens)
+        self.register_buffer("indices", kept_indices + block_mask.size(0) * offsets)
+
+    def compute_slice(self, num_steps: int, prev_steps: int = 0) -> torch.Tensor:
+        total_steps = num_steps + prev_steps
+        num_blocks = math.ceil(total_steps / self.block_size)
+        indices = self.indices[: num_blocks * self.num_kept_tokens]
+        return indices[torch.logical_and(prev_steps <= indices, indices < total_steps)] - prev_steps
+
+    def forward(self, *args, **kwargs):
+        raise NotImplementedError
+
+
+class Head(Slicer):
+    def __init__(self, max_blocks: int, block_mask: torch.Tensor, head_module: nn.Module) -> None:
+        super().__init__(max_blocks, block_mask)
+        assert isinstance(head_module, nn.Module)
+        self.head_module = head_module
+
+    def forward(
+        self, x: torch.Tensor, output_hidden_states: bool = False, num_steps: int = None, prev_steps: int = None
+    ) -> torch.Tensor:
+        hidden_states = () if output_hidden_states else None
+        x_sliced = x[:, self.compute_slice(num_steps, prev_steps)]  # x is (B, T, E)
+        # Add the hidden state to the tuple after Relu activation
+        x_sliced_hidden = self.head_module[1](self.head_module[0](x_sliced))
+        hidden_states = hidden_states + (x_sliced_hidden,) if output_hidden_states else None
+        hidden_states = hidden_states + (self.head_module[2](x_sliced_hidden),) if output_hidden_states else None
+
+        return self.head_module(x_sliced), hidden_states
+
+
+class Embedder(nn.Module):
+    def __init__(self, max_blocks: int, block_masks: List[torch.Tensor], embedding_tables: List[nn.Embedding]) -> None:
+        super().__init__()
+        assert len(block_masks) == len(embedding_tables)
+        assert (sum(block_masks) == 1).all()  # block mask are a partition of a block
+        self.embedding_dim = embedding_tables[0].embedding_dim
+        assert all(e.embedding_dim == self.embedding_dim for e in embedding_tables)
+        self.embedding_tables = embedding_tables
+        self.slicers = [Slicer(max_blocks, block_mask) for block_mask in block_masks]
+
+    def forward(self, tokens: torch.Tensor, num_steps: int, prev_steps: int) -> torch.Tensor:
+        assert tokens.ndim == 2  # x is (B, T)
+        output = torch.zeros(*tokens.size(), self.embedding_dim, device=tokens.device)
+        for slicer, emb in zip(self.slicers, self.embedding_tables):
+            s = slicer.compute_slice(num_steps, prev_steps)
+            output[:, s] = emb(tokens[:, s])
+        return output
+
+
+@dataclass
+class EncoderDecoderConfig:
+    resolution: int
+    in_channels: int
+    z_channels: int
+    ch: int
+    ch_mult: List[int]
+    num_res_blocks: int
+    attn_resolutions: List[int]
+    out_ch: int
+    dropout: float
+
+
+class Encoder(nn.Module):
+    def __init__(self, config: EncoderDecoderConfig) -> None:
```

Comment:
If a component of a model has its own config, it should be defined in the model's corresponding config module - `configuration_iris` not in the modeling file

Reply:
:+1: :100: moving it